RecoMuon/MuonRechitClusterProducer/plugins/RechitClusterProducer.cc

```diff
@@ -0,0 +1,268 @@
+  const edm::ESGetToken<typename Trait::GeometryType, MuonGeometryRecord> GeometryToken_;
+  edm::EDGetTokenT<typename Trait::InputType> InputToken_;
+  typedef std::vector<reco::MuonRecHitCluster> RecHitClusterCollection;
+
+  const double rParam_;      // distance paramter
+  const int nRechitMin_;     // min number of rechits
+  const int nStationThres_;  // min number of rechits to count towards nStation
+};
+
+template <typename Trait>
+RechitClusterProducerT<Trait>::RechitClusterProducerT(const edm::ParameterSet& iConfig)
+    : GeometryToken_(esConsumes<typename Trait::GeometryType, MuonGeometryRecord>()),
+      InputToken_(consumes<typename Trait::InputType>(iConfig.getParameter<edm::InputTag>("recHitLabel"))),
+      rParam_(iConfig.getParameter<double>("rParam")),
+      nRechitMin_(iConfig.getParameter<int>("nRechitMin")),
+      nStationThres_(iConfig.getParameter<int>("nStationThres")) {
+  produces<RecHitClusterCollection>();
+}
+
+//
+// member functions
+//
+
+// ------------ method called to produce the data  ------------
+template <typename Trait>
+void RechitClusterProducerT<Trait>::produce(edm::StreamID, edm::Event& ev, const edm::EventSetup& iSetup) const {
+  auto const& geo = iSetup.getData(GeometryToken_);
+  auto const& rechits = ev.get(InputToken_);
+
+  std::vector<fastjet::PseudoJet> fjInput;
+  RecHitRefVector inputs_;
+
+  fastjet::JetDefinition jet_def(fastjet::cambridge_algorithm, rParam_);
+
+  inputs_.clear();
+  fjInput.clear();
+
+  Trait RecHitTrait;
+
+  int recIt = 0;
+  for (auto const& rechit : rechits) {
+    LocalPoint RecHitLocalPosition = rechit.localPosition();
+    auto detid = RecHitTrait.detid(rechit);
+    auto thischamber = geo.chamber(detid);
+    if (thischamber) {
+      GlobalPoint globalPosition = thischamber->toGlobal(RecHitLocalPosition);
+      float x = globalPosition.x();
+      float y = globalPosition.y();
+      float z = globalPosition.z();
+      RechitRef ref = RechitRef(&rechits, recIt);
+      inputs_.push_back(ref);
+      fjInput.push_back(fastjet::PseudoJet(x, y, z, globalPosition.mag()));
+      fjInput.back().set_user_index(recIt);
+    }
+    recIt++;
+  }
+  fastjet::ClusterSequence clus_seq(fjInput, jet_def);
+
+  //keep all the clusters
+  double ptmin = 0.0;
+  std::vector<fastjet::PseudoJet> fjJets = fastjet::sorted_by_pt(clus_seq.inclusive_jets(ptmin));
+
+  auto clusters = std::make_unique<RecHitClusterCollection>();
+  if (!fjJets.empty()) {
+    for (unsigned int ijet = 0; ijet < fjJets.size(); ++ijet) {
+      // get the fastjet jet
+      const fastjet::PseudoJet& fjJet = fjJets[ijet];
+
+      // skip if the cluster has too few rechits
+      if (int(fjJet.constituents().size()) < nRechitMin_)
+        continue;
+      // get the constituents from fastjet
+      RecHitRefVector rechits;
+      for (unsigned int i = 0; i < fjJet.constituents().size(); i++) {
+        auto index = fjJet.constituents()[i].user_index();
```

Comment:
```suggestion
      for (auto const &constituent: fjJet.constituents()) {
        auto index = constituent.user_index();
```
the `vector<PseudoJet> PseudoJet::constituents()` doesn't look like a trivial access method; need to avoid re-extraction of the full vector

Reply:
done